Source/JavaScriptCore/runtime/WaiterListManager.h

```diff
@@ -0,0 +1,139 @@
+#include <wtf/ParkingLot.h>
+#include <wtf/RawPointer.h>
+
+namespace JSC {
+
+class WaiterListManager {
+    WTF_MAKE_FAST_ALLOCATED;
+
+public:
+    static WaiterListManager& singleton();
+
+    void addWaiter(void* ptr, JSPromise* promise, Seconds timeout)
+    {
+        RELEASE_ASSERT(ptr && promise);
+        VM& vm = promise->globalObject()->vm();
+        vm.apiLock().lock();
+        auto ticket = vm.deferredWorkTimer->addPendingWork(vm, promise, { });
+        vm.apiLock().unlock();
+
+        ParkingLot::parkAsync(ptr, ticket);
+
+        m_registeredVMs.add(&vm);
+        RunLoop::current().dispatchAfter(timeout, [ptr, ticket] {
+            WaiterListManager::singleton().timeout(ptr, ticket);
+        });
+    }
+
+    void timeout(void* ptr, DeferredWorkTimer::Ticket ticket)
+    {
+        RELEASE_ASSERT(ptr && ticket);
+        JSPromise* promise = jsCast<JSPromise*>(ticket->target());
+        VM* vm = &promise->globalObject()->vm();
+        if (!m_registeredVMs.contains(vm))
+            return;
+
+        ParkingLot::unparkAsync(ptr, ticket, [this](void* data) {
+            bool isOKString = false;
+            notifyWaiterImpl(data, isOKString);
+        });
+    }
+
+    unsigned notifyWaiter(void* ptr, unsigned count)
+    {
+        RELEASE_ASSERT(ptr);
+        bool isAsyncMode = true;
+        unsigned notified = ParkingLot::unparkCount(
+            ptr, count, [this](void* data) {
+                bool isOKString = true;
+                notifyWaiterImpl(data, isOKString);
+            },
+            isAsyncMode);
+        return notified;
+    }
+
+    void notifyWaiterImpl(void* data, const bool isOKString)
+    {
+        auto ticket = static_cast<DeferredWorkTimer::Ticket>(data);
+        VM* vm = &jsCast<JSPromise*>(ticket->target())->globalObject()->vm();
+        vm->deferredWorkTimer->scheduleWorkSoon(ticket, [isOKString](DeferredWorkTimer::Ticket ticket) {
+            JSPromise* promise = jsCast<JSPromise*>(ticket->target());
+            JSGlobalObject* globalObject = promise->globalObject();
+            VM& vm = promise->globalObject()->vm();
+            promise->resolve(globalObject, (isOKString ? vm.smallStrings.okString() : vm.smallStrings.timedOutString()));
+        });
+    }
+
+    void unregister(VM* vm)
+    {
+        m_registeredVMs.remove(vm);
+    }
+
+    size_t waiterListSize(void* ptr)
+    {
+        return ParkingLot::bucketSize(ptr);
+    }
+
+private:
+    struct RegisteredVMs {
+        WTF_MAKE_FAST_ALLOCATED;
+
+    public:
+        void add(VM* vm)
+        {
+            LockHolder locker(lock);
+            set.add(vm);
+        }
+
+        void remove(VM* vm)
+        {
+            LockHolder locker(lock);
+            set.remove(vm);
+        }
+
+        bool contains(VM* vm)
+        {
+            LockHolder locker(lock);
+            return set.contains(vm);
+        }
+
+    private:
+        mutable Lock lock;
+        HashSet<VM*> set;
+    };
```

Comment:
I don't think this is good abstraction. We should just carefully manage Lock and HashSet in `WaiterListManager`.